File: packages/i18nify-go/modules/currency/currency_formatters_test.go
package currency

import (
	"testing"
)

// ── groupIntegerStr unit tests ────────────────────────────────────────────────

func TestGroupIntegerStr(t *testing.T) {
	cases := []struct {
		name       string
		input      string
		groupSizes []int
		sep        string
		want       string
	}{
		{"western [3]", "1234567", []int{3}, ",", "1,234,567"},
		{"indian [3,2] 7-digit", "1234567", []int{3, 2}, ",", "12,34,567"},
		{"indian [3,2] 8-digit", "12345678", []int{3, 2}, ",", "1,23,45,678"},
		{"empty groupSizes", "1234567", []int{}, ",", "1234567"},
		{"empty sep", "1234567", []int{3}, "", "1234567"},
		{"single group", "567", []int{3}, ",", "567"},
		{"shorter than group", "12", []int{3}, ",", "12"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := groupIntegerStr(c.input, c.groupSizes, c.sep)
			if got != c.want {
				t.Errorf("got %q, want %q", got, c.want)
			}
		})
	}
}

// ── substituteDigitsStr unit tests ───────────────────────────────────────────

func TestSubstituteDigitsStr(t *testing.T) {
	arabicDigits := "٠١٢٣٤٥٦٧٨٩"
	bengaliDigits := "০১২৩৪৫৬৭৮৯"

	cases := []struct {
		name   string
		input  string
		glyphs string
		want   string
	}{
		{"arabic digits", "1234567", arabicDigits, "١٢٣٤٥٦٧"},
		{"bengali digits", "1234567", bengaliDigits, "১২৩৪৫৬৭"},
		{"separators pass through", "1,234.56", arabicDigits, "١,٢٣٤.٥٦"},
		{"wrong glyph length", "123", "abc", "123"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := substituteDigitsStr(c.input, c.glyphs)
			if got != c.want {
				t.Errorf("got %q, want %q", got, c.want)
			}
		})
	}
}

// ── CurrencyFormatters constant ───────────────────────────────────────────────

func TestCurrencyFormattersMap(t *testing.T) {
	styles := []FormatterStyle{
		StyleWestern, StyleIndian, StyleEuropean, StyleFrench, StyleSwiss,
		StyleSwazi, StyleArabic, StyleBengali, StyleNoGrouping,
	}
	if len(CurrencyFormatters) != len(styles) {
		t.Errorf("expected %d styles, got %d", len(styles), len(CurrencyFormatters))
	}
	for _, s := range styles {
		if _, ok := CurrencyFormatters[s]; !ok {
			t.Errorf("missing style %q", s)
		}
	}

	if CurrencyFormatters[StyleIndian].GroupSizes[0] != 3 || CurrencyFormatters[StyleIndian].GroupSizes[1] != 2 {
		t.Error("INDIAN should have GroupSizes [3,2]")
	}
	if CurrencyFormatters[StyleArabic].Digits != "٠١٢٣٤٥٦٧٨٩" {
		t.Error("ARABIC missing digit glyphs")
	}
	if CurrencyFormatters[StyleNoGrouping].ThousandsSep != "" {
		t.Error("NO_GROUPING should have empty ThousandsSep")
	}
}

// ── FormatWithStyle integration tests ────────────────────────────────────────

func TestFormatWithStyle(t *testing.T) {
	type args struct {
		amount float64
		style  FormatterStyle
		opts   FormatWithStyleOptions
	}

	cases := []struct {
		name string
		args args
		want string
	}{
		{"WESTERN basic", args{1234567.89, StyleWestern, FormatWithStyleOptions{Decimals: 2}}, "1,234,567.89"},
		{"INDIAN basic", args{1234567.89, StyleIndian, FormatWithStyleOptions{Decimals: 2}}, "12,34,567.89"},
		{"EUROPEAN basic", args{1234567.89, StyleEuropean, FormatWithStyleOptions{Decimals: 2}}, "1.234.567,89"},
		{"FRENCH basic", args{1234567.89, StyleFrench, FormatWithStyleOptions{Decimals: 2}}, "1 234 567,89"},
		{"SWISS basic", args{1234567.89, StyleSwiss, FormatWithStyleOptions{Decimals: 2}}, "1'234'567.89"},
		{"SWAZI basic", args{1234567.89, StyleSwazi, FormatWithStyleOptions{Decimals: 2}}, "1 234 567.89"},
		{"ARABIC basic", args{1234567.89, StyleArabic, FormatWithStyleOptions{Decimals: 2}}, "١٬٢٣٤٬٥٦٧٫٨٩"},
		{"BENGALI basic", args{1234567.89, StyleBengali, FormatWithStyleOptions{Decimals: 2}}, "১২,৩৪,৫৬৭.৮৯"},
		{"NO_GROUPING basic", args{1234567.89, StyleNoGrouping, FormatWithStyleOptions{Decimals: 2}}, "1234567.89"},

		// Currency-aware decimal places
		{"INR 2 decimals", args{1234.5, StyleIndian, FormatWithStyleOptions{Currency: "INR", Decimals: -1}}, "1,234.50"},
		{"JPY 0 decimals", args{1234.9, StyleWestern, FormatWithStyleOptions{Currency: "JPY", Decimals: -1}}, "1,235"},
		{"BHD 3 decimals", args{1.5, StyleWestern, FormatWithStyleOptions{Currency: "BHD", Decimals: -1}}, "1.500"},

		// Negative
		{"WESTERN negative", args{-1234.56, StyleWestern, FormatWithStyleOptions{Decimals: 2}}, "-1,234.56"},
		{"EUROPEAN negative", args{-1234.56, StyleEuropean, FormatWithStyleOptions{Decimals: 2}}, "-1.234,56"},

		// Decimals override
		{"3 decimal places", args{1234.567, StyleWestern, FormatWithStyleOptions{Decimals: 3}}, "1,234.567"},
		{"0 decimal places", args{1234.567, StyleWestern, FormatWithStyleOptions{Decimals: 0}}, "1,235"},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got, err := FormatWithStyle(c.args.amount, c.args.style, c.args.opts)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got != c.want {
				t.Errorf("got %q, want %q", got, c.want)
			}
		})
	}
}

func TestFormatWithStyleShowSymbol(t *testing.T) {
	got, err := FormatWithStyle(1000, StyleIndian, FormatWithStyleOptions{
		Currency:   "INR",
		Decimals:   -1,
		ShowSymbol: true,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != "₹1,000.00" {
		t.Errorf("got %q, want %q", got, "₹1,000.00")
	}
}

func TestFormatWithStyleUnknownStyle(t *testing.T) {
	_, err := FormatWithStyle(100, "INVALID_STYLE", FormatWithStyleOptions{Decimals: 2})
	if err == nil {
		t.Error("expected error for unknown style, got nil")
	}
}

File: packages/i18nify-js/src/modules/currency/__tests__/currencyFormatters.test.ts
import {
  CURRENCY_FORMATTERS,
  groupIntegerPart,
  substituteDigits,
} from '../currencyFormatters';
import { formatWithStyle } from '../index';
import { CurrencyCodeType } from '../types';

// ── groupIntegerPart unit tests ──────────────────────────────────────────────

describe('groupIntegerPart', () => {
  it('applies Western [3] grouping', () => {
    expect(groupIntegerPart('1234567', [3], ',')).toBe('1,234,567');
  });

  it('applies Indian [3,2] grouping', () => {
    expect(groupIntegerPart('1234567', [3, 2], ',')).toBe('12,34,567');
    expect(groupIntegerPart('12345678', [3, 2], ',')).toBe('1,23,45,678');
  });

  it('returns the string unchanged when groupSizes is empty', () => {
    expect(groupIntegerPart('1234567', [], ',')).toBe('1234567');
  });

  it('returns the string unchanged when sep is empty', () => {
    expect(groupIntegerPart('1234567', [3], '')).toBe('1234567');
  });

  it('handles a single-group number', () => {
    expect(groupIntegerPart('567', [3], ',')).toBe('567');
  });

  it('handles a number shorter than the first group', () => {
    expect(groupIntegerPart('12', [3], ',')).toBe('12');
  });
});

// ── substituteDigits unit tests ──────────────────────────────────────────────

describe('substituteDigits', () => {
  const ARABIC_DIGITS = '٠١٢٣٤٥٦٧٨٩';
  const BENGALI_DIGITS = '০১২৩৪৫৬৭৮৯';

  it('substitutes Arabic-Indic digits', () => {
    expect(substituteDigits('1234567', ARABIC_DIGITS)).toBe('١٢٣٤٥٦٧');
  });

  it('substitutes Bengali digits', () => {
    expect(substituteDigits('1234567', BENGALI_DIGITS)).toBe('১২৩৪৫৬৭');
  });

  it('leaves non-digit characters (commas, dots) unchanged', () => {
    // substituteDigits only maps 0-9 → locale glyphs; separators pass through as-is
    expect(substituteDigits('1,234.56', ARABIC_DIGITS)).toBe('١,٢٣٤.٥٦');
  });
});

// ── CURRENCY_FORMATTERS constant ─────────────────────────────────────────────

describe('CURRENCY_FORMATTERS', () => {
  it('exposes all 9 styles', () => {
    const expected = [
      'WESTERN',
      'INDIAN',
      'EUROPEAN',
      'FRENCH',
      'SWISS',
      'SWAZI',
      'ARABIC',
      'BENGALI',
      'NO_GROUPING',
    ];
    expect(Object.keys(CURRENCY_FORMATTERS)).toEqual(expected);
  });

  it('WESTERN uses comma thousands and period decimal', () => {
    expect(CURRENCY_FORMATTERS.WESTERN.thousandsSep).toBe(',');
    expect(CURRENCY_FORMATTERS.WESTERN.decimalSep).toBe('.');
    expect(CURRENCY_FORMATTERS.WESTERN.groupSizes).toEqual([3]);
  });

  it('INDIAN uses [3,2] grouping', () => {
    expect(CURRENCY_FORMATTERS.INDIAN.groupSizes).toEqual([3, 2]);
  });

  it('EUROPEAN uses period thousands and comma decimal', () => {
    expect(CURRENCY_FORMATTERS.EUROPEAN.thousandsSep).toBe('.');
    expect(CURRENCY_FORMATTERS.EUROPEAN.decimalSep).toBe(',');
  });

  it('FRENCH uses narrow no-break space thousands', () => {
    expect(CURRENCY_FORMATTERS.FRENCH.thousandsSep).toBe(' ');
    expect(CURRENCY_FORMATTERS.FRENCH.decimalSep).toBe(',');
  });

  it('SWISS uses apostrophe thousands', () => {
    expect(CURRENCY_FORMATTERS.SWISS.thousandsSep).toBe("'");
    expect(CURRENCY_FORMATTERS.SWISS.decimalSep).toBe('.');
  });

  it('ARABIC has digit glyph substitution', () => {
    expect(CURRENCY_FORMATTERS.ARABIC.digits).toBe('٠١٢٣٤٥٦٧٨٩');
    expect(CURRENCY_FORMATTERS.ARABIC.thousandsSep).toBe('٬');
    expect(CURRENCY_FORMATTERS.ARABIC.decimalSep).toBe('٫');
  });

  it('BENGALI uses Indian [3,2] grouping with Bengali digits', () => {
    expect(CURRENCY_FORMATTERS.BENGALI.groupSizes).toEqual([3, 2]);
    expect(CURRENCY_FORMATTERS.BENGALI.digits).toBe('০১২৩৪৫৬৭৮৯');
  });

  it('NO_GROUPING has empty thousandsSep and empty groupSizes', () => {
    expect(CURRENCY_FORMATTERS.NO_GROUPING.thousandsSep).toBe('');
    expect(CURRENCY_FORMATTERS.NO_GROUPING.groupSizes).toEqual([]);
  });
});

// ── formatWithStyle integration tests ────────────────────────────────────────

describe('currency - formatWithStyle', () => {
  const amount = 1234567.89;

  it('WESTERN: 1,234,567.89', () => {
    expect(formatWithStyle(amount, 'WESTERN')).toBe('1,234,567.89');
  });

  it('INDIAN: 12,34,567.89', () => {
    expect(formatWithStyle(amount, 'INDIAN')).toBe('12,34,567.89');
  });

  it('EUROPEAN: 1.234.567,89', () => {
    expect(formatWithStyle(amount, 'EUROPEAN')).toBe('1.234.567,89');
  });

  it('FRENCH: 1 234 567,89', () => {
    expect(formatWithStyle(amount, 'FRENCH')).toBe('1 234 567,89');
  });

  it("SWISS: 1'234'567.89", () => {
    expect(formatWithStyle(amount, 'SWISS')).toBe("1'234'567.89");
  });

  it('SWAZI: 1 234 567.89 (space separator)', () => {
    expect(formatWithStyle(amount, 'SWAZI')).toBe('1 234 567.89');
  });

  it('ARABIC: uses Arabic-Indic digits and separators', () => {
    const result = formatWithStyle(amount, 'ARABIC');
    expect(result).toBe('١٬٢٣٤٬٥٦٧٫٨٩');
  });

  it('BENGALI: uses Bengali digits with Indian grouping', () => {
    const result = formatWithStyle(amount, 'BENGALI');
    expect(result).toBe('১২,৩৪,৫৬৭.৮৯');
  });

  it('NO_GROUPING: 1234567.89', () => {
    expect(formatWithStyle(amount, 'NO_GROUPING')).toBe('1234567.89');
  });

  // Currency-aware decimal places
  it('respects currency minor_unit for INR (2 decimals)', () => {
    expect(
      formatWithStyle(1234.5, 'INDIAN', {
        currency: 'INR' as CurrencyCodeType,
      }),
    ).toBe('1,234.50');
  });

  it('respects currency minor_unit for JPY (0 decimals)', () => {
    expect(
      formatWithStyle(1234.9, 'WESTERN', {
        currency: 'JPY' as CurrencyCodeType,
      }),
    ).toBe('1,235');
  });

  it('respects currency minor_unit for BHD (3 decimals)', () => {
    expect(
      formatWithStyle(1.5, 'WESTERN', { currency: 'BHD' as CurrencyCodeType }),
    ).toBe('1.500');
  });

  // showSymbol option
  it('prepends INR symbol when showSymbol=true', () => {
    const result = formatWithStyle(1000, 'INDIAN', {
      currency: 'INR' as CurrencyCodeType,
      showSymbol: true,
    });
    expect(result).toBe('₹1,000.00');
  });

  // decimals override
  it('overrides decimal places via options.decimals', () => {
    expect(formatWithStyle(1234.567, 'WESTERN', { decimals: 3 })).toBe(
      '1,234.567',
    );
    expect(formatWithStyle(1234.567, 'WESTERN', { decimals: 0 })).toBe('1,235');
  });

  // Negative amounts
  it('preserves negative sign', () => {
    expect(formatWithStyle(-1234.56, 'WESTERN')).toBe('-1,234.56');
    expect(formatWithStyle(-1234.56, 'EUROPEAN')).toBe('-1.234,56');
  });

  // Error cases
  it('throws for NaN', () => {
    expect(() => formatWithStyle(NaN, 'WESTERN')).toThrow(
      "'amount' must be a finite number",
    );
  });

  it('throws for Infinity', () => {
    expect(() => formatWithStyle(Infinity, 'WESTERN')).toThrow(
      "'amount' must be a finite number",
    );
  });
});
